Readme.md

# koa-ip-filter

 IP blacklist middleware for koa 2 (Only koa 2 supported)

## Installation

```js
$ npm install git+ssh://git@github.com/ahauser31/koa-ip-filter.git
```

## Example

```js
var ipFilter = require('koa-ip-filter');
var redis = require('redis');
var koa = require('koa');
var app = new Koa();

// apply ip filter

app.use(ipFilter({
  db: redis.createClient(),
  id: function (context) {
    return context.ip;
  }
}));

// response middleware

app.use(function (ctx, next){
  ctx.body = 'Hello!';
});

app.listen(3000);
console.log('listening on port 3000');
```

## Options
 
 - `db` redis connection instance
 - `ip` ip to compare requests (can be changed to use value other than client ip) [`(ctx) => (return ctx.ip)`]
 - `duration` blacklist duration in milliseconds [86400000 (=24 hours)]
 - `errorMsgPermanent` text in body of error response for permanent blacklisting [`'Permanently blacklisted'`]
 - `errorMsgRetry` text in body of error response for limited blacklisting (limited for `duration`) [`'Blacklisted, retry in '`]
 - `appendRetryTime` append retry time to error body text [true]
 - `errorMsgDB` message of thrown error object in case of DB errors [`'ERROR_DATABASE_ERROR'`]
 - `customDBError` thrown error message for DB failure is `errorMsgDB` instead of redis error message (hide true cause in production) [true]
 - `retryHeader` custom header for non-permanent blacklisting [`'X-Retry-After'`]
 - `setHeader` append custom header to response for non-permanent blacklisting [true]
 - `filterBlacklist` error message thrown downstream indicating that client should be non-permanently blacklisted for `duration` [`'IP_FILTER_BLACKLIST'`]
 - `filterBlacklistPermanent` error message thrown downstream indicating that client should be permanently blacklisted [`'IP_FILTER_BLACKLIST_PERMANENT'`]
 - `throw` throw for client ip found on blacklist  [true]
 - `log` logging function to receive error message [`(ctx, msg) => ()`]

## Responses

  Example 200 with header fields:

```
HTTP/1.1 200 OK
X-Powered-By: koa
Content-Type: text/plain; charset=utf-8
Content-Length: 6
Date: Wed, 13 Nov 2013 21:22:13 GMT
Connection: keep-alive

Hello!
```

  Example 401 response:

```
HTTP/1.1 401 Unauthorized
X-Powered-By: koa
Content-Type: text/plain; charset=utf-8
Content-Length: 39
X-Retry-After: 60000
Date: Wed, 13 Nov 2013 21:21:48 GMT
Connection: keep-alive

Blacklisted, retry in 1 minute
```

## License

  MIT


History.md
1.0.0 / 2016-08-15
==================

 * initial release


test/ipfilter.js

var request = require('supertest');
var should = require('should');
var redis = require('redis');
var Koa = require('koa');

var ip_filter = require('..');

var db = redis.createClient();

describe('ip filtering middleware', function() {

  before(function(done) {
    db.keys('ipFilter:*', function(err, rows) {
      rows.forEach( (element) => {
        db.del(element, () => {});  
      });
    });

    done();
  });

  describe('blacklist', function() {
    var app;
    var yielded;

    beforeEach(function(done) {
      app = new Koa();

      app.use(ip_filter({
        db: db,
        errorMsgRetry: 'Blacklisted, non-permanent',
        errorMsgPermanent: 'Permanently blacklisted',
        appendRetryTime: false,
        throw: false,
        ip: function(ctx) {return '127.0.0.1'}
      }));
  
      yielded = false;
      
      app.use(function (ctx, next) {
        yielded = true;
        ctx.body = 'Not blocked!';
      });
      
      done();
    });
    
    it('responds with 401 when IP is non-permanently blacklisted', function(done) {
      db.set('ipFilter:127.0.0.1', Date.now() + 86400000, (err, tReply) => {
        if (err) throw('Database error');
        
        request(app.listen())
          .get('/')
          .expect( (res) => {
            res.status.should.equal(401);
            res.text.should.equal('Blacklisted, non-permanent');
          })
          .end(done);
      });
    });
    
    it('responds with 403 when IP is permanently blacklisted', function(done) {
      db.set('ipFilter:127.0.0.1', -1, (err, tReply) => {
        if (err) throw('Database error');
        
        request(app.listen())
          .get('/')
          .expect( (res) => {
            res.status.should.equal(403);
            res.text.should.equal('Permanently blacklisted');
          })
          .end(done);
      });
    });

    it('should not yield downstream if IP is blacklisted', function(done) {
      db.set('ipFilter:127.0.0.1', -1, (err, tReply) => {
        if (err) throw('Database error');

        request(app.listen())
          .get('/')
          .expect((res) => {
            res.status.should.equal(403);
            yielded.should.equal(false);
          })
          .end(done);
      });
    });
    
    it('should yield downstream if IP is not blacklisted', function(done) {
      db.del('ipFilter:127.0.0.1', (err, tReply) => {
        if (err) throw('Database error');

        request(app.listen())
          .get('/')
          .expect((res) => {
            res.status.should.equal(200);
            res.text.should.equal('Not blocked!');
            yielded.should.equal(true);
          })
          .end(done);
      });
    });
    
  });
  
  describe('app puts ip on blacklist, permanent and wait for duration', function() {
    var app;
    var block;

    before(function(done) {
      app = new Koa();
      block = 0;

      app.use(ip_filter({
        db: db,
        duration: 1000,
        errorMsgRetry: 'Blacklisted, non-permanent',
        errorMsgPermanent: 'Permanently blacklisted',
        appendRetryTime: false,
        filterBlacklist: 'IP_FILTER_BLACKLIST',
        filterBlacklistPermanent: 'IP_FILTER_BLACKLIST_PERMANENT',
        retryHeader: 'X-Retry-After',
        throw: false,
        ip: function(ctx) {return '127.0.0.1'}
      }));
      
      app.use( (ctx, next) => {
        if (block === 0) return next();
        if (block < 0) ctx.throw(500, 'IP_FILTER_BLACKLIST_PERMANENT');
        if (block > 0) ctx.throw(500, 'IP_FILTER_BLACKLIST');
      });
      
      app.use(function (ctx, next) {
        ctx.body = 'Not blocked!';
      });
      
      db.del('ipFilter:127.0.0.1', (err, tReply) => {
        if (err) throw('Database error');
        done();
      });
    });
  
    it('should yield downstream', function(done) {
      block = 0;
      request(app.listen())
        .get('/')
        .expect((res) => {
          res.status.should.equal(200);
          res.text.should.equal('Not blocked!');
        })
        .end(done);
    });
    
    it('should cause the app to be blacklisted non-permanently, as an error is thrown downstream', function(done) {
      block = -1;
      request(app.listen())
        .get('/')
        .expect(403, 'Permanently blacklisted')
        .end(done);
    });
    
    it('should wait 1500ms and check if client still blacklisted (longer than non-permablock duration)', function(done) {
      block = 0;
      setTimeout( () => {
        request(app.listen())
          .get('/')
          .expect(403, 'Permanently blacklisted')
          .end(done);
      }, 1500);
    });
    
  });
  
  describe('app puts ip on blacklist, non-permanent and wait for duration', function() {
    var app;
    var block;

    before(function(done) {
      app = new Koa();
      block = 0;

      app.use(ip_filter({
        db: db,
        duration: 1000,
        errorMsgRetry: 'Blacklisted, non-permanent',
        errorMsgPermanent: 'Permanently blacklisted',
        appendRetryTime: false,
        filterBlacklist: 'IP_FILTER_BLACKLIST',
        filterBlacklistPermanent: 'IP_FILTER_BLACKLIST_PERMANENT',
        retryHeader: 'X-Retry-After',
        throw: false,
        ip: function(ctx) {return '127.0.0.1'}
      }));
      
      app.use( (ctx, next) => {
        if (block === 0) return next();
        if (block < 0) ctx.throw(500, 'IP_FILTER_BLACKLIST_PERMANENT');
        if (block > 0) ctx.throw(500, 'IP_FILTER_BLACKLIST');
      });
      
      app.use(function (ctx, next) {
        ctx.body = 'Not blocked!';
      });
      
      db.del('ipFilter:127.0.0.1', (err, tReply) => {
        if (err) throw('Database error');
        done();
      });
    });
    
    it('should yield downstream', function(done) {
      block = 0;
      request(app.listen())
        .get('/')
        .expect((res) => {
          res.status.should.equal(200);
          res.text.should.equal('Not blocked!');
        })
        .end(done);
    });
    
    it('should cause the app to be blacklisted non-permanently, as an error is thrown downstream', function(done) {
      block = 1;
      request(app.listen())
        .get('/')
        .expect(401, 'Blacklisted, non-permanent')
        .end(done);
    });
    
    it('should wait 500ms and check if client still blacklisted (shorter than blacklist duration)', function(done) {
      block = 0;
      setTimeout( () => {
        request(app.listen())
          .get('/')
          .expect(function(res) {
            res.status.should.equal(401);
            res.text.should.equal('Blacklisted, non-permanent');
            res.headers.should.containEql('x-retry-after');
            res.headers['x-retry-after'].should.be.below(1000);
          })
          .end(done);
      }, 500);
    });
    
    it('should wait another 1000ms and check if client unblocked again (longer than blacklist duration)', function(done) {
      block = 0;
      setTimeout( () => {
        request(app.listen())
          .get('/')
          .expect(200, 'Not blocked!')
          .end(done);
      }, 1000);
    });

  });
  
  describe('blacklist with throw', function() {
    var app;

    beforeEach(function(done) {
      app = new Koa();

      app.use(function (ctx, next) {
        return next().catch(err => {
          ctx.body = err.message;
          if (typeof err.headers !== 'undefined') ctx.set(err.headers);
        });
      });
      
      app.use(ip_filter({
        db: db,
        errorMsgRetry: 'Blacklisted, non-permanent',
        errorMsgPermanent: 'Permanently blacklisted',
        appendRetryTime: false,
        throw: true,
        ip: function(ctx) {return '127.0.0.1'}
      }));
  
      
      app.use(function (ctx, next) {
        ctx.body = 'Not blocked';
      });
      
      done();
    });

    it('responds with 401 when IP is non-permanently blacklisted', function(done) {
      db.set('ipFilter:127.0.0.1', Date.now() + 86400000, (err, tReply) => {
        if (err) throw('Database error');
        
        request(app.listen())
          .get('/')
          .expect( (res) => {
            res.status.should.equal(401);
            res.text.should.equal('Blacklisted, non-permanent');
          })
          .end(done);
      });
    });
    
    it('responds with 403 when IP is permanently blacklisted', function(done) {
      db.set('ipFilter:127.0.0.1', -1, (err, tReply) => {
        if (err) throw('Database error');
        
        request(app.listen())
          .get('/')
          .expect( (res) => {
            res.status.should.equal(403);
            res.text.should.equal('Permanently blacklisted');
          })
          .end(done);
      });
    });
    
  });

  
  describe('ip', function (done) {
    it('should allow specifying a custom `ip` function', function (done) {
      db.set('ipFilter:bar', -1, (err, tReply) => {
        var app = new Koa();
    
        app.use(ip_filter({
          db: db,
          throw: false,
          errorMsgPermanent: 'Permanently blacklisted',
          ip: function (ctx) {
            return ctx.request.header.foo;
          }
        }));
        
        app.use( (ctx, next) => {
          ctx.body = 'Not blacklisted!';
        });
    
        request(app.listen())
          .get('/')
          .set('foo', 'bar')
          .expect(function(res) {
            res.text.should.equal('Permanently blacklisted');
          })
          .end(done);
      });
    });
    
    it('should not block if `ip` returns `false`', function (done) {
      db.set('ipFilter:bar', -1, (err, tReply) => {
        var app = new Koa();
    
        app.use(ip_filter({
          db: db,
          throw: false,
          errorMsgPermanent: 'Permanently blacklisted',
          ip: function (ctx) {
            return false;
          }
        }));
        
        app.use( (ctx, next) => {
          ctx.body = 'Not blacklisted!';
        });
    
        request(app.listen())
          .get('/')
          .set('foo', 'bar')
          .expect(function(res) {
            res.text.should.equal('Not blacklisted!');
          })
          .end(done);
      });
    });
  
    it('should block using the `ip` value', function (done) {
      var app = new Koa();
  
      app.use(ip_filter({
        db: db,
        throw: false,
        ip: function (ctx) {
          return ctx.request.header.foo;
        }
      }));
  
      app.use(function (ctx, next) {
        ctx.body = ctx.request.header.foo;
      });
  
      request(app.listen())
        .get('/')
        .set('foo', 'bar')
        .expect(200, 'bar')
        .end(function() {
          request(app.listen())
            .get('/')
            .set('foo', 'biz')
            .expect(200, 'biz')
            .end(done);
        });
    });
  });
  
  describe('custom error message in body', function() {
    var app;

    beforeEach(function(done) {
      app = new Koa();

      app.use(ip_filter({
        db: db,
        errorMsgRetry: 'Boo!',
        errorMsgPermanent: 'Hoo!',
        appendRetryTime: false,
        throw: false,
        ip: function(ctx) {return '127.0.0.1'}
      }));
  
      
      app.use(function (ctx, next) {
        ctx.body = 'Not blocked!';
      });
      
      done();
    });
    
    it('body text of reply is "Boo!" when IP is non-permanently blacklisted', function(done) {
      db.set('ipFilter:127.0.0.1', Date.now() + 86400000, (err, tReply) => {
        if (err) throw('Database error');
        
        request(app.listen())
          .get('/')
          .expect( (res) => {
            res.status.should.equal(401);
            res.text.should.equal('Boo!');
          })
          .end(done);
      });
    });
    
    it('body text of reply is "Hoo!" when IP is permanently blacklisted', function(done) {
      db.set('ipFilter:127.0.0.1', -1, (err, tReply) => {
        if (err) throw('Database error');
        
        request(app.listen())
          .get('/')
          .expect( (res) => {
            res.status.should.equal(403);
            res.text.should.equal('Hoo!');
          })
          .end(done);
      });
    });
    
  });
  
  describe('reply header', function() {
    var app;

    beforeEach(function(done) {
      app = new Koa();

      app.use(ip_filter({
        db: db,
        retryHeader: 'Foo',
        setHeader: true,
        appendRetryTime: false,
        throw: false,
        ip: function(ctx) {return '127.0.0.1'}
      }));
  
      
      app.use(function (ctx, next) {
        ctx.body = 'Not blocked!';
      });
      
      done();
    });
    
    it('headers of reply contains custom retry header "Foo" when IP is non-permanently blacklisted', function(done) {
      db.set('ipFilter:127.0.0.1', Date.now() + 86400000, (err, tReply) => {
        if (err) throw('Database error');
        
        request(app.listen())
          .get('/')
          .expect( (res) => {
            res.status.should.equal(401);
            res.headers.should.containEql('foo');
          })
          .end(done);
      });
    });
    
    it('retry header contains less than "duration" milliseconds', function(done) {
      db.set('ipFilter:127.0.0.1', Date.now() + 1000, (err, tReply) => {
        if (err) throw('Database error');
        
        request(app.listen())
          .get('/')
          .expect( (res) => {
            res.status.should.equal(401);
            res.headers.should.containEql('foo');
            res.headers['foo'].should.be.belowOrEqual(1000);
          })
          .end(done);
      });
    });
    
  });
  
  describe('append retry time', function() {
    var app;

    beforeEach(function(done) {
      app = new Koa();

      app.use(ip_filter({
        db: db,
        setHeader: true,
        appendRetryTime: true,
        throw: false,
        ip: function(ctx) {return '127.0.0.1'}
      }));
  
      
      app.use(function (ctx, next) {
        ctx.body = 'Not blocked!';
      });
      
      done();
    });
    
    
    it('body of reply contains retry time', function(done) {
      db.set('ipFilter:127.0.0.1', Date.now() + 500, (err, tReply) => {
        if (err) throw('Database error');
        
        request(app.listen())
          .get('/')
          .expect( (res) => {
            res.status.should.equal(401);
            res.headers.should.containEql('x-retry-after');
            res.text.should.endWith(res.headers['x-retry-after'] + ' ms');
          })
          .end(done);
      });
    });
    
  });
  
  describe('logging function', function() {
    var app, logMsg, logStatus;

    beforeEach(function(done) {
      app = new Koa();
      logMsg = null;
      logStatus = 500;

      app.use(ip_filter({
        db: db,
        errorMsgPermanent: 'ERROR_BLACKLIST_PERMANENT',
        appendRetryTime: false,
        throw: false,
        ip: function(ctx) {return '127.0.0.1'},
        log: function(ctx, msg) { logMsg = msg; logStatus = ctx.status; }
      }));
  
      
      app.use(function (ctx, next) {
        ctx.body = 'Not blocked!';
      });
      
      done();
    });

    it('should log out correct error message', function(done) {
      db.set('ipFilter:127.0.0.1', -1, (err, tReply) => {
        if (err) throw('Database error');
        
        request(app.listen())
          .get('/')
          .expect( (res) => {
            res.status.should.equal(403);
            logMsg.should.equal('ERROR_BLACKLIST_PERMANENT');
            logStatus.should.equal(403);
          })
          .end(done);
      });
    });
    
  });

});


index.js

/**
 * Module dependencies.
 */
var ms = require('ms');
var assert = require('assert');

/**
 * Expose `ip_filter()`.
 */

module.exports = ip_filter;

/**
 * Initialize ip filer middleware with the given `opts`:
 *
 * - `db` redis connection instance
 * - `id` id to compare requests [ (ctx) => (return ctx.ip) ]
 * - `duration` blacklist duration in milliseconds [24 hour]
 * - `errorMsgPermanent` text in body of error response for permanent blacklisting ['Permanently blacklisted']
 * - `errorMsgRetry` text in body of error response for limited blacklisting (limited for `duration`) ['Blacklisted, retry in ']
 * - `appendRetryTime` append retry time to error body text [true]
 * - `errorMsgDB` message of thrown error object in case of DB errors ['ERROR_DATABASE_ERROR']
 * - `customDBError` thrown error message for DB failure is `errorMsgDB` instead of redis error message (hide true cause in production) [true]
 * - `retryHeader` custom header for non-permanent blacklisting ['X-Retry-After']
 * - `setHeader` append custom header to response for non-permanent blacklisting [true]
 * - `filterBlacklist` error message thrown downstream indicating that client should be non-permanently blacklisted for `duration` ['IP_FILTER_BLACKLIST']
 * - `filterBlacklistPermanent` error message thrown downstream indicating that client should be permanently blacklisted ['IP_FILTER_BLACKLIST_PERMANENT']
 * - `throw` throw for client ip found on blacklist  [true]
 * - `log` logging function to receive error message [(ctx, msg) => ()]
 *
 * @param {Object} opts
 * @return {Function}
 * @api public
 */

function ip_filter(opts) {
  opts = opts || {};
  assert(opts.db, '.db required');
  opts.duration = opts.duration || 86400000;  // 24h
  opts.errorMsgPermanent = opts.errorMsgPermanent || 'Permanently blacklisted';
  opts.errorMsgRetry = opts.errorMsgRetry || 'Blacklisted, retry in ';
  opts.appendRetryTime = (typeof opts.appendRetryTime !== 'undefined') ? opts.appendRetryTime : true;
  opts.errorMsgDB = opts.errorMsgDB || 'ERROR_DATABASE_ERROR';
  opts.customDBError = (typeof opts.customDBError !== 'undefined') ? opts.customDBError : true;
  opts.retryHeader = opts.retryHeader || 'X-Retry-After';
  opts.setHeader = (typeof opts.setHeader !== 'undefined') ? opts.setHeader : true;
  
  opts.filterBlacklist = opts.filterBlacklist || 'IP_FILTER_BLACKLIST';
  opts.filterBlacklistPermanent = opts.filterBlacklistPermanent || 'IP_FILTER_BLACKLIST_PERMANENT';
  opts.throw = (typeof opts.throw !== 'undefined') ? opts.throw : false;

  return function (ctx, next){
    // Ignore request if no IP present
    var ip = opts.ip ? opts.ip(ctx) : ctx.ip;
    if (false === ip) return next();

    // initialize
    var entry = 'ipFilter:' + ip;

    // Check db if ip is on blacklist
    return new Promise( (resolve, reject) => {
      opts.db.get(entry, function (err, res) {
        if (err) return reject(opts.customDBError ? opts.errorMsgDB : err); // No need to write the reject handler, Koa throws anyway
        
        // Return time of ban or 0 for no ban (-1 == premanent)
        return resolve(res === null ? 0 : res);
      });
    }).then((res) => {
      if (res !== 0)
      {
        // Client on blacklist, return an error (no need to update DB, it will clean old blacklisting automatically after `duration`)
        error(res, false);
        return;
      } else {
        // Not on blacklist
        return next().catch( (err) => {
          // Caught a downstream error, check if it is an error indicating the client needs to be put on blacklist
          if (err.message === opts.filterBlacklist || err.message === opts.filterBlacklistPermanent ) {
            // Client to be blacklisted, return error and write entry into DB
            error(err.message === opts.filterBlacklistPermanent ? -1 : opts.duration, true);
            return;
          }
          else
          {
            // Re-throw error for upstream handlers
            ctx.throw(err);
            return;
          }
        });
      }
    });
    
    function error(res, writeDB)
    {
      // Save IP into DB
      if (writeDB)
      {
        var val = [entry, (res < 0 ? res : (Date.now() + res))];
        if (res >= 0) {
          val.push('PX');
          val.push(res);
        }
        val.push('NX');
        opts.db.set(val, (err, res) => {
          // Throw if value can't be written
          if (err) ctx.throw(500, opts.customDBError ? opts.errorMsgDB : err);
        });
      }
      
      // IP on blacklist
      if (res < 0)
      {
        // Permanent ban
        ctx.status = 403;
        ctx.body = opts.errorMsgPermanent;
        
        if (opts.log) opts.log(ctx, opts.errorMsgPermanent);
        
        if (opts.throw) ctx.throw(ctx.status, ctx.body);
        return; 
      }
      
      // Non-permanent ban, calculate time left to unban
      var retryTime = res - Date.now();
      var headers = {};
      headers[opts.retryHeader] = retryTime;
      if (opts.setHeader) ctx.set(headers);
      
      ctx.status = 401;
      ctx.body = opts.errorMsgRetry + (opts.appendRetryTime ? ms(retryTime, { long: true }) : '');
      
      if (opts.log) opts.log(ctx, opts.errorMsgRetry);
      
      if (opts.throw) ctx.throw(ctx.status, ctx.body, { headers: opts.setHeader ? headers : [] });
    }
    
  }
}
